{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "170adc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pyBigWig\n",
    "import pyranges as pr\n",
    "import genomicranges as gr\n",
    "from matplotlib import pyplot as plt\n",
    "import glob \n",
    "\n",
    "sns.set_context(\"talk\")\n",
    "\n",
    "chr_categories = [str(i) for i in list(range(1,23))]\n",
    "chr_categories.extend([\"X\", \"Y\"])\n",
    "chr_categories = [\"chr\"+i for i in chr_categories]\n",
    "\n",
    "GENOME_LENGHT = pr.data.chromsizes().End.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "240d183a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_intervals(intervals):\n",
    "    # Sort the intervals by their starting point\n",
    "    sorted_intervals = sorted(intervals, key=lambda x: x[0])\n",
    "\n",
    "    # Initialize the merged interval list with the first interval\n",
    "    merged_intervals = [sorted_intervals[0]]\n",
    "\n",
    "    # Iterate over the sorted intervals, merging overlapping intervals\n",
    "    for interval in sorted_intervals[1:]:\n",
    "        last_merged = merged_intervals[-1]\n",
    "\n",
    "        # If the current interval overlaps with the last merged interval, merge them\n",
    "        if interval[0] <= last_merged[1]:\n",
    "            merged_intervals[-1] = (last_merged[0], max(last_merged[1], interval[1]))\n",
    "        else:\n",
    "            merged_intervals.append(interval)\n",
    "\n",
    "    # Calculate the total length of the merged intervals\n",
    "    total_length = sum(end - start for start, end in merged_intervals)\n",
    "\n",
    "    return merged_intervals, total_length"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b085518c",
   "metadata": {},
   "source": [
    "## Check how much of the region is covered with binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5f16aa2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Collect stats \n",
    "n_peaks = pd.DataFrame()\n",
    "chromosome = \"chr1\"\n",
    "\n",
    "def get_chr_stats(chromosome, bbb):\n",
    "    len_chr = bbb.chroms()[chromosome]\n",
    "    peaks = bbb.entries(chromosome, 0,len_chr,withString=False )\n",
    "\n",
    "    # Create peak ranges object\n",
    "    peaks_ranges = pd.DataFrame(peaks)\n",
    "    peaks_ranges.columns = [\"start\", \"end\"]\n",
    "    peaks_ranges[\"len\"] = peaks_ranges.start - peaks_ranges.end\n",
    "    # Stats \n",
    "    merged_intervals, total_length = merge_intervals(peaks)\n",
    "\n",
    "\n",
    "    # Create Summary DF\n",
    "    entry = pd.DataFrame([{\"chr\": chromosome, \"n_peaks\": len(peaks), \"len\": len_chr, \"covered_len\": total_length}])\n",
    "    entry[\"perc_covered\"] = (entry.covered_len * 100) / entry.len\n",
    "    bed_stats = pd.concat([n_peaks,entry],ignore_index=True)\n",
    "    return(bed_stats)\n",
    "\n",
    "def get_bed_stats(bbb): \n",
    "    bed_stats = pd.DataFrame()\n",
    "    for chromosome in set(bbb.chroms().keys()): \n",
    "            bed_stats = pd.concat([bed_stats, get_chr_stats(chromosome, bbb)])\n",
    "    return(bed_stats)\n",
    "\n",
    "def prep_bed_stats_chrom(bb):\n",
    "    bbb = pyBigWig.open(bb)\n",
    "    tissue = bb.split(\"/\")[-2]\n",
    "    filename = bb.split(\"/\")[-1]\n",
    "    bed_stats = get_bed_stats(bbb)\n",
    "    bed_stats[\"chr\"] = pd.Categorical(bed_stats[\"chr\"], chr_categories)\n",
    "    bed_stats = bed_stats.sort_values(by = \"chr\")\n",
    "    bed_stats[\"tissue\"] = tissue\n",
    "    bed_stats[\"filename\"] = filename\n",
    "    return(bed_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc154725",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pool_bed_stats(bb):\n",
    "    bbb = pyBigWig.open(bb)\n",
    "    tissue = bb.split(\"/\")[-2]\n",
    "    filename = bb.split(\"/\")[-1]\n",
    "    bed_stats = get_bed_stats_pooled(bbb)\n",
    "    bed_stats[\"tissue\"] = tissue\n",
    "    bed_stats[\"filename\"] = filename\n",
    "    return(bed_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21807da8",
   "metadata": {},
   "source": [
    "## Which is the coverage of the genome?\n",
    "### Across tissues? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "459e8b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "files = glob.glob(\"/home/luisasantus/Desktop/crg_cluster/data/FED/out/downloads/Homo sapiens/ATAC-seq/*/*.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c7fb9879",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bed_pooled_coverage(bed): \n",
    "    tissue = bed.split(\"/\")[-2]\n",
    "    filename = bed.split(\"/\")[-1]\n",
    "    covered =  pr.read_bed(bed).merge(strand=False).length\n",
    "    entry = pd.DataFrame([{\"tissue\": tissue, \"filename\": filename, \"covered\": covered, \"perc_covered\": covered*100/GENOME_LENGHT}])\n",
    "    return(entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "477bcecf",
   "metadata": {},
   "outputs": [
    {
     "ename": "ParserError",
     "evalue": "Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mParserError\u001b[0m                               Traceback (most recent call last)",
      "Input \u001b[0;32mIn [14]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m summary_beds_coverage \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat(\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mmap\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mget_bed_pooled_coverage\u001b[49m\u001b[43m,\u001b[49m\u001b[43mfiles\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m, ignore_index\u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "Input \u001b[0;32mIn [13]\u001b[0m, in \u001b[0;36mget_bed_pooled_coverage\u001b[0;34m(bed)\u001b[0m\n\u001b[1;32m      2\u001b[0m tissue \u001b[38;5;241m=\u001b[39m bed\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m      3\u001b[0m filename \u001b[38;5;241m=\u001b[39m bed\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m----> 4\u001b[0m covered \u001b[38;5;241m=\u001b[39m  \u001b[43mpr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_bed\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbed\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mmerge(strand\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\u001b[38;5;241m.\u001b[39mlength\n\u001b[1;32m      5\u001b[0m entry \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame([{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtissue\u001b[39m\u001b[38;5;124m\"\u001b[39m: tissue, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfilename\u001b[39m\u001b[38;5;124m\"\u001b[39m: filename, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcovered\u001b[39m\u001b[38;5;124m\"\u001b[39m: covered, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mperc_covered\u001b[39m\u001b[38;5;124m\"\u001b[39m: covered\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m\u001b[38;5;241m/\u001b[39mGENOME_LENGHT}])\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m(entry)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pyranges/readers.py:89\u001b[0m, in \u001b[0;36mread_bed\u001b[0;34m(f, as_df, nrows)\u001b[0m\n\u001b[1;32m     86\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m:\n\u001b[1;32m     87\u001b[0m     header \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m---> 89\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     90\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     91\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\n\u001b[1;32m     92\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mChromosome\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcategory\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     93\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mStrand\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcategory\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m     94\u001b[0m \u001b[43m    \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     95\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnrows\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnrows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     96\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheader\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     97\u001b[0m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     99\u001b[0m df\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m columns[:df\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m]]\n\u001b[1;32m    101\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m as_df:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py:680\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    665\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    666\u001b[0m     dialect,\n\u001b[1;32m    667\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    676\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m    677\u001b[0m )\n\u001b[1;32m    678\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 680\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py:581\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    578\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[1;32m    580\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[0;32m--> 581\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1254\u001b[0m, in \u001b[0;36mTextFileReader.read\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1252\u001b[0m nrows \u001b[38;5;241m=\u001b[39m validate_integer(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnrows\u001b[39m\u001b[38;5;124m\"\u001b[39m, nrows)\n\u001b[1;32m   1253\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1254\u001b[0m     index, columns, col_dict \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1255\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1256\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py:225\u001b[0m, in \u001b[0;36mCParserWrapper.read\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m    223\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    224\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlow_memory:\n\u001b[0;32m--> 225\u001b[0m         chunks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reader\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    226\u001b[0m         \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n\u001b[1;32m    227\u001b[0m         data \u001b[38;5;241m=\u001b[39m _concatenate_chunks(chunks)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/_libs/parsers.pyx:805\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.read_low_memory\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/_libs/parsers.pyx:861\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/_libs/parsers.pyx:847\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/site-packages/pandas/_libs/parsers.pyx:1960\u001b[0m, in \u001b[0;36mpandas._libs.parsers.raise_parser_error\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mParserError\u001b[0m: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'."
     ]
    }
   ],
   "source": [
    "summary_beds_coverage = pd.concat(list(map(get_bed_pooled_coverage,files)), ignore_index= True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbbe0435",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.displot(summary_beds_coverage, x = \"perc_covered\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cafc217",
   "metadata": {},
   "source": [
    "### And what if we group all of them? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a0994de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "collapsed_atac = \"/home/luisasantus/Desktop/crg_cluster/data/FED/out/downloads/Homo sapiens/ATAC-seq/collapsed/ATAC-seq.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "776a9b2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>filename</th>\n",
       "      <th>covered</th>\n",
       "      <th>perc_covered</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>collapsed</td>\n",
       "      <td>ATAC-seq.bed</td>\n",
       "      <td>283811469</td>\n",
       "      <td>9.167943</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tissue      filename    covered  perc_covered\n",
       "0  collapsed  ATAC-seq.bed  283811469      9.167943"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_bed_pooled_coverage(collapsed_atac)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1dbc06b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ad380a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1ab1c0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c60f7d5a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3750ae47",
   "metadata": {},
   "source": [
    "## Paralogs analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2eaec3d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_ids = \"/home/luisasantus/Desktop/crg_cluster/data/FED/gene_lists/dgd_Hsa_all_v71.genes\"\n",
    "id_df = pd.read_csv(list_ids, sep = \"\\t\", header = None)\n",
    "ids = list(id_df[0])\n",
    "\n",
    "test_genes = \"/home/luisasantus/Desktop/crg_cluster/data/FED/gene_lists/dgd_Hsa_all_v71.tsv\"\n",
    "gene_groups = pd.read_csv(test_genes, sep = \"\\t\")\n",
    "#gene_pairs = gene_groups[gene_groups.NB_Genes > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "12fe44f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We are working now with the pooled one\n",
    "enrichment_result =  \"/home/luisasantus/Desktop/crg_cluster/data/FED/out/enrichment/dgd_Hsa_all_v71_ENCFF012SCX_0.50_overlap_binary.txt\"\n",
    "enrichment = pd.read_csv(enrichment_result, sep = \"\\t\", header = None)\n",
    "enrichment.columns = [\"ENS_ID\", \"track_binary\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6067c0fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here I can have different comparison functions\n",
    "def comparison_binary_equality(df):\n",
    "    column_index = df.columns.get_loc(\"track_binary\")\n",
    "    a = df.to_numpy() \n",
    "    return(a[0] == a).all(0)[column_index]\n",
    "\n",
    "def get_distance(df):\n",
    "    end_index = df.columns.get_loc(\"end\")\n",
    "    start_index = df.columns.get_loc(\"start\")\n",
    "    max_end = max(df[\"end\"])\n",
    "    min_end = min(df[\"start\"])\n",
    "    return max_end - min_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "af2f02e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add enrichment tracks to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "22524f18",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pairs_enriched = gene_groups.merge(enrichment, on = \"ENS_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fe0e470b",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_track_comparison = pd.DataFrame(gene_pairs_enriched.groupby([\"group_id\"]).apply(comparison_binary_equality)).reset_index()\n",
    "group_track_comparison = group_track_comparison.rename(columns= {0:\"track_binary_eval\"})\n",
    "gene_pairs_eval = gene_pairs_enriched.merge(group_track_comparison, on = \"group_id\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f2052f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many were called open \n",
    "summary_genes_ocr = gene_pairs_eval.track_binary.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b18b6e10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paralog genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>closed</th>\n",
       "      <td>3451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>open</th>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        paralog genes\n",
       "closed           3451\n",
       "open               46"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(summary_genes_ocr).rename(index = { 0 : \"closed\", 1: \"open\"}, columns = {\"track_binary\": \"paralog genes\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "686dad65",
   "metadata": {},
   "source": [
    "# Check the distances across paralogs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c7e2657d",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_distance = pd.DataFrame(gene_pairs_enriched.groupby([\"group_id\"]).apply(get_distance)).reset_index()\n",
    "group_distance = group_distance.rename(columns= {0:\"distance\"})\n",
    "gene_pairs_eval_distance = gene_pairs_eval.merge(group_distance, on = \"group_id\" )\n",
    "chr_count = gene_pairs_eval_distance.groupby(\"group_id\")[\"chr\"].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b884d20f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3451\n",
       "1      46\n",
       "Name: track_binary, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pairs_eval_distance.track_binary.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "860ce443",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>group_id</th>\n",
       "      <th>NB_Genes</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ENS_ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>track_binary</th>\n",
       "      <th>track_binary_eval</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1385069</td>\n",
       "      <td>1405538</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000215915</td>\n",
       "      <td>ATAD3C</td>\n",
       "      <td>ATPase family, AAA domain containing 3C [Sourc...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>84998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1407143</td>\n",
       "      <td>1433228</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>ATPase family, AAA domain containing 3B [Sourc...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>84998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1447531</td>\n",
       "      <td>1470067</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000197785</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>ATPase family, AAA domain containing 3A [Sourc...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>84998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1570603</td>\n",
       "      <td>1590473</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000248333</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>cyclin-dependent kinase 11B [Source:HGNC Symbo...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>85163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1634169</td>\n",
       "      <td>1655766</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000008128</td>\n",
       "      <td>CDK11A</td>\n",
       "      <td>cyclin-dependent kinase 11A [Source:HGNC Symbo...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>85163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3492</th>\n",
       "      <td>Y</td>\n",
       "      <td>1010</td>\n",
       "      <td>4</td>\n",
       "      <td>25365594</td>\n",
       "      <td>25437503</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000205944</td>\n",
       "      <td>DAZ2</td>\n",
       "      <td>deleted in azoospermia 2 [Source:HGNC Symbol;A...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>1777681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3493</th>\n",
       "      <td>Y</td>\n",
       "      <td>1010</td>\n",
       "      <td>4</td>\n",
       "      <td>26909216</td>\n",
       "      <td>26959626</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000187191</td>\n",
       "      <td>DAZ3</td>\n",
       "      <td>deleted in azoospermia 3 [Source:HGNC Symbol;A...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>1777681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3494</th>\n",
       "      <td>Y</td>\n",
       "      <td>1010</td>\n",
       "      <td>4</td>\n",
       "      <td>26980008</td>\n",
       "      <td>27053183</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000205916</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>deleted in azoospermia 4 [Source:HGNC Symbol;A...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>1777681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3495</th>\n",
       "      <td>Y</td>\n",
       "      <td>1011</td>\n",
       "      <td>2</td>\n",
       "      <td>26191376</td>\n",
       "      <td>26194166</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000172352</td>\n",
       "      <td>CDY1B</td>\n",
       "      <td>chromodomain protein, Y-linked, 1B [Source:HGN...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>1579673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3496</th>\n",
       "      <td>Y</td>\n",
       "      <td>1011</td>\n",
       "      <td>2</td>\n",
       "      <td>27768264</td>\n",
       "      <td>27771049</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>chromodomain protein, Y-linked, 1 [Source:HGNC...</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>1579673</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3497 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  group_id  NB_Genes     start       end  strand           ENS_ID  \\\n",
       "0      1         1         3   1385069   1405538       1  ENSG00000215915   \n",
       "1      1         1         3   1407143   1433228       1  ENSG00000160072   \n",
       "2      1         1         3   1447531   1470067       1  ENSG00000197785   \n",
       "3      1         2         2   1570603   1590473      -1  ENSG00000248333   \n",
       "4      1         2         2   1634169   1655766      -1  ENSG00000008128   \n",
       "...   ..       ...       ...       ...       ...     ...              ...   \n",
       "3492   Y      1010         4  25365594  25437503       1  ENSG00000205944   \n",
       "3493   Y      1010         4  26909216  26959626      -1  ENSG00000187191   \n",
       "3494   Y      1010         4  26980008  27053183       1  ENSG00000205916   \n",
       "3495   Y      1011         2  26191376  26194166      -1  ENSG00000172352   \n",
       "3496   Y      1011         2  27768264  27771049       1  ENSG00000172288   \n",
       "\n",
       "        Name                                        Description  track_binary  \\\n",
       "0     ATAD3C  ATPase family, AAA domain containing 3C [Sourc...             0   \n",
       "1     ATAD3B  ATPase family, AAA domain containing 3B [Sourc...             0   \n",
       "2     ATAD3A  ATPase family, AAA domain containing 3A [Sourc...             0   \n",
       "3     CDK11B  cyclin-dependent kinase 11B [Source:HGNC Symbo...             0   \n",
       "4     CDK11A  cyclin-dependent kinase 11A [Source:HGNC Symbo...             0   \n",
       "...      ...                                                ...           ...   \n",
       "3492    DAZ2  deleted in azoospermia 2 [Source:HGNC Symbol;A...             0   \n",
       "3493    DAZ3  deleted in azoospermia 3 [Source:HGNC Symbol;A...             0   \n",
       "3494    DAZ4  deleted in azoospermia 4 [Source:HGNC Symbol;A...             0   \n",
       "3495   CDY1B  chromodomain protein, Y-linked, 1B [Source:HGN...             0   \n",
       "3496    CDY1  chromodomain protein, Y-linked, 1 [Source:HGNC...             0   \n",
       "\n",
       "      track_binary_eval  distance  \n",
       "0                  True     84998  \n",
       "1                  True     84998  \n",
       "2                  True     84998  \n",
       "3                  True     85163  \n",
       "4                  True     85163  \n",
       "...                 ...       ...  \n",
       "3492               True   1777681  \n",
       "3493               True   1777681  \n",
       "3494               True   1777681  \n",
       "3495               True   1579673  \n",
       "3496               True   1579673  \n",
       "\n",
       "[3497 rows x 12 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pairs_eval_distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "45595074",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAakAAAEZCAYAAAAt5touAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAiUUlEQVR4nO3debgdVZ3u8e8bFKEJ5JgwSlTEIIhwtcGoeImEyatIIAQRFGWybYfGFmwFmsbLcIEGbG4j3osNijK0zaQkTAFBSWS4HQ2CgRDmIVcQRBJyEiAgJL/+Y9UOlco+ezh777Nrn/N+nqeefbJq1apVu07qd1atVbUUEZiZmZXRqG5XwMzMbCAOUmZmVloOUmZmVloOUmZmVloOUmZmVlpv6nYFhhNJr5MC/9Ju18XMrIdsAKyMiDVikjwEvX0krQQ0ZsyYblfFzKxn9Pf3A0RErHF3zy2p9lo6ZsyYMUuWLOl2PczMekZfXx/9/f1V70C5T8rMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLz0mVxCuvvMLcuXPXSJ84cSLrrLNOF2pkZtZ9DlIlMXfuXI46bwZ94yesSlvy1KOc8zWYNGlSF2tmZtY9DlIl0jd+Ahtt9YFuV8PMrDTcJ2VmZqXlIGVmZqXlIGVmZqXlIGVmZqXlIGVmZqXlIGVmZqXlIGVmZqXV1SAlabyk70m6Q9KLkkLS5EKeDSSdIOnXkv6U5Zsn6WhJaxfyTs7KqLZsU2X/e0qaI2m5pOcknS+pr6MHbWZmDet2S2oC8FngReBXA+R5B/AN4HfAl4B9gOuAM4ErB9jmWGCnwvJkPkMWDGcCfwCmAN/Kyr5BUre/FzMzo/tvnLgtIjYGkDSVFCSKngC2iIiXcmm3SnoNOEnS9hFxX2GbhyNiTp19nwXMBw6MiJVZHZ4BbgYOAK5o+mjMzKytutpiqASHOnleKgSoisrbWMc3u19JmwMTgUvzdYiIW4Cngf2bLdPMzNqvl29r7QYEsKDKuvMlvS6pX9L1knYsrN8u+5xfZdv7cutXI2lJrQUYM8hjMTOzKnoySEn6EPB1UktoYW5VP3AO8LfArsC3gW2BOyV9OJdvXPa5uErxi3Przcysi7rdJ9U0SROAa4EHSYFqlYi4B7gnl3S7pGtJLabTgD0KxcUAu6maHhF9deq2BLemzMzapqdaUpK2BGYBLwB7RsTSettExLOkwRAfySUvyj6rtZjGUr2FZWZmQ6xngpSkd5EC1CvA7hHxXBObj2L11tH92We1vqftqd5XZWZmQ6wngpSkd5IC1Apgt4j4YxPbbgrsCawakh4RTwF3AQfnn4mStDuwOXB1m6puZmYt6HqflKRPZz9OzD53kbQh8FJE3ChpY+BWYGPgCGDzbAh5xWMR8eesrJ8CjwN3k24JbkN6sHdd4B8Luz6WdBvwMkkXAG8jPSD8G+Cq9h6lmZkNRteDFGsGhJOyz4XAFqTReVtmaZdV2f5w4KLs5/uAg0gDKtYj9T3NBk6NiNVu4UXErZL2Bk4GbgCWATOAYyJixSCPxczM2qjrQSoiVGf9bKBmnlzeM4Azmtj3TcBNjeY3M7Oh1RN9UmZmNjI5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWk5SJmZWWl1NUhJGi/pe5LukPSipJA0eYC8e0qaI2m5pOcknS+pr0q+0ZLOlfRMlvcuSfu0UqaZmXVHt1tSE4DPAi8CvxooUxa4ZgJ/AKYA3wL2AW6QVDyG6cDBwAnAp4AFwHRJe7VQppmZdcGburz/2yJiYwBJU0lBopqzgPnAgRGxMsv/DHAzcABwRZa2F7AHMC0ipmdps4AtgbNJQampMs3MrHu62mKoBIdaJG0OTAQuzeePiFuAp4H9c9n3A/qBa3L5ArgY2EbStoMo08zMuqQXbmttl33Or7Luvtz6St4FVYLfvYWyminTzMy6pNu3+xoxLvtcXGXdYmCHQt6HB8iXL6uZMleRtKRWRYExddabmVkTeqElVRENpg+Ur5m8tcowM7Mh0gstqUXZ57gq68ayemtoUY185PI2U+YqEdFXq6JZS8utKTOzNumFltT92We1fqLtWb1f6X7gvVWGkG+ffc7P5Wu0TDMz65LSB6mIeAq4Czg4H3wk7Q5sDlydyz4d6CM995R3CPBQRCwYRJlmZtYlXb/dJ+nT2Y8Ts89dJG0IvBQRN2Zpx5KeX7pM0gXA24Azgd8AV+WKmwnMAi6UNA54AjgU2BnYt7DrRss0M7Mu6XqQYs2AcFL2uRDYAiAibpW0N3AycAOwDJgBHBMRKyobRkRkDwWfni19pDdOTIuI6/I7abRMMzPrnq4HqYhQg/luAm5qIN9S4MhsaUuZZmbWHaXvkzIzs5HLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzErLQcrMzEqrqSAl6XFJ+9RYv7ekx1uvlpmZWfMtqS2A0TXWrwe8c9C1MTMzy2n37b5NgJfbXKaZmY1QdaePl/QxYHIuaZqkCVWyjgUOAn7flpqZmdmIVzdIAbsCJ2Y/BzAtW6p5FDi6DfUyMzNrKEidA1wECHgcOAq4ppAngBcjYnEb62ZmZiNc3SAVEf1AP4CkXYEHIuK5TlfMzMyskZbUKhHx605VxMzMrKipIAUg6R3Al4GtgHGk24B5ERG7t6FuZmY2wjUVpCR9EpgOrA0sA9wHZWZmHdNsS+qfgeeBqRFxVwfqY2ZmtkqzD/NuA5wz1AFK0kWSosayaZZv9gDrL69S5mhJ50p6RtJySXfVeuWTmZkNvWZbUn8G/tKJitTxv4B/K6S9GfgFcG9EPJtLfwQ4pJD3+SplTgd2AI4BngAOA6ZLmhIRM9tRaTMza02zQepSYH/g3A7UZUAR8RjwWD5N0jRgXeDCQvaXI2JOrfIk7QXsAUyLiOlZ2ixgS+BswEHKzKwEmr3ddxGwtqRrJO0m6V2S3lFcOlDPao4gvSfwikFsux/p2a9VDyVHRAAXA9tI2rYtNTQzs5Y025J6kPR2CQF718i31qBr1ABJmwGfAH4aEUsLq7eW9AKwPuk23sXAmRHxWi7PdsCCiFhZ2Pbe/Poq+11Sp2pjGjsCMzNrRLNB6hRSkOq2Q0mBsHir73bgclIwHQ1MJdV5R1LrqWIc8HCVchfn1puZWZc1+8aJkzpUj2YdBjwaEbflEyPiO4V810v6E3C8pJ0j4o589hrlV10XEX21KpW1tNyaMjNrk56bPl7SzsDWwE8a3OTi7HOnXNoiqreWxmaffkjZzKwEmn3jxMcayVds4bTZEcAK3gg+9VQCcb7/6X5gf0mjCv1S22ef81uropmZtUOzfVKzaaxPqiMDJyStBxwA/CIinm5ws8ozU/lh6dOBLwJTWH3akUOAhyJijUETZmY29JoNUocPUMa7Sf1ETwLnt1almg4kDYj4cXGFpEnAccDPgYXAesC+pDpfFRF35rLPBGYBF0oaRxoFeCiwc7aNmZmVQLMDJwa8xSbpu8DdLdeotsNJb4+4tsq6Z7LPU4ANSbf3HgK+CXw/nzEiQtJU4PRs6SMNOZ8WEdd1ouJmZta8pqfqGEhEvCDpR6TXDDXaX9TsPibVWPco8KkmyloKHJktZmZWQu0e3fcC6dVCZmZmLWtbkJK0DvAF4Nl6ec3MzBrR7BD0NQYsZMaSnkPaCPh2q5UyMzOD5vukDhsgfTHpNUNHR8R/tFQjMzOzTLOj+3ruDRVmZta7HHTMzKy0BjUEXdIGpEkDKyP5HgduiYhl7aqYmZlZ00FK0t+QZq8dTZpXCtKrkl6U9M2IKE6fYWZmNijNju7bB7iA1HL6n7zxItb3AV8HLpD0nN/aYGZm7dBsS+oY4AHgwxHxYi79V5J+QnqJ67GAg5SZmbWs2YET7wcuKgQoALL+qIuzPGZmZi0bzOg+1VhXhqnlzcxsmGg2SM0DDs3mdVqNpNGkh33ntaFeZmZmTfdJ/QtwNXC3pHNJ01vAGwMnJgDT2lc9MzMbyZp948QMSUcCZ5LmaKrc3hPwEnBkRFwz0PZmZmbNaPo5qYg4T9J/AHsC7yIFqMdID/P2t7l+ZmY2gg3qjRMRsQS4qr1VMTMzW13dgROS1pJ0hqSv1Mn3VUmnS6o1+s/MzKxhjYzu+zxpjqi5dfL9lvQg72dbrZSZmRk0FqQ+A/wyIn5XK1O2/hc4SJmZWZs0EqR2BH7ZYHmzgA8OvjpmZmZvaCRIjQWea7C8P2f5zczMWtZIkFoGbNhgeeOANd7rZ2ZmNhiNBKn7gY83WN6eWX4zM7OWNRKkrgb2kLRvrUzZXFN7Aj9vR8XMzMwaCVLnA48CV0o6TdIW+ZWStpB0KnAl8HCWv60kTZYUAyzbFPLuKWmOpOWSnpN0vqS+KmWOlnSupGeyvHdlgdbMzEqi7hsnImK5pE8B1wP/CBwnaRmwFFgf2ID0aqSHgL0j4pUO1vdY4LZC2pOVHyRNBmYCM4ATgLeR3jO4naRJEbEyt910YAfSRI5PkN7gPl3SlIiY2ZHam5lZUxp6LVJEPCrpA8CXgE+T3nq+KSlQ3U66xfejiFjeoXpWPBwRc2qsP4s0pf2BlYAk6RngZuAA4IosbS9gD2BaREzP0mYBWwJnkwKdmZl1WcPzSUXEKxHx/YjYJSI2jIi1s8/JWXqnA1RNkjYHJgKX5ltMEXEL8DSwfy77fkA/cE0uX5BmFt5G0rZDUmkzM6tpMDPzdtP5kl6X1C/pekk75tZtl33Or7Ldfbn1lbwLCrf/AO4tlGVmZl00qLegd0E/cA4wG1gMvBc4DrhT0i4R8RvSM1pk64sWk/qfKsaRBnlUy1dZvwZJS+rUc0yd9WZm1oSeCFIRcQ9wTy7pdknXklpNp5H6l1ZlH6iYOv9udJ2ZmQ2RnghS1UTEs5JuBirDxhdln9VaQWNZvYW1qEY+qN4aIyL6atUpa2m5NWVm1ia91idVNIo3Wj2VN11U60/antX7qu4H3iupePzbZ5/V+rXMzGyI9WyQkrQp6Q0XcwAi4ingLuDgfPCRtDuwOenNGRXTgT5gSqHYQ4CHImJB52puZmaN6onbfZJ+CjwO3A28AGxDerB3XdIDxhXHkp6JukzSBbzxMO9vWH26+5mkaUUulDSO9DDvocDOQM3XP5mZ2dDpiSBFGkJ+EPB1YD1Sn9Js4NSIWHVrLiJulbQ3cDJwA+kN7jOAYyJiRS5fSJoKnJ4tfcAC0sO913X+cBqz4vXXmDdv3mppEydOZJ111ulSjczMhlZPBKmIOAM4o8G8NwE3NZBvKXBktpTSsmcXct4Ty9ls4VoALHnqUc75GkyaNKnLNTMzGxo9EaRGsvU325KNtvpAt6thZtYVPTtwwszMhj8HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzK62eCFKSdpd0kaSHJL0s6SlJV0vavpBvtqSoslxepczRks6V9Iyk5ZLukrTP0B2VmZnV86ZuV6BBXwHGAf8KPABsAhwDzJU0OSLm5PI+AhxS2P75KmVOB3bIynkCOAyYLmlKRMxsb/XNzGwweiVI/V1EPJdPkHQzKbh8G9g/t+rlQtBag6S9gD2AaRExPUubBWwJnA04SJmZlUBP3O4rBqgsbQmp1TR+EEXuB/QD1+TKC+BiYBtJ2w6upmZm1k49EaSqkbQRsB0wv7Bqa0kvSHpd0iOSTpD05kKe7YAFEbGykH5vbr2ZmXVZr9zuW40kAReQguy/5FbdDlwOPAiMBqYCpwA7klpPFeOAh6sUvTi3vtp+l9Sp2pg6683MrAk9GaSA75IC0OER8UAlMSK+U8h3vaQ/AcdL2jki7sitixrl11pnZmZDpOeClKTTgH8AvhERFzWwycXA8cBOQCVILaJ6a2ls9rm4yjoioq9O3ZbQwdbUitdfY968eWukT5w4kXXWWadTuzUz65qeClKSTiEFnGMi4twGN6v0u+X7n+4H9pc0qtAvVXnuqtjPVQrLnl3IeU8sZ7OFa61KW/LUo5zzNZg0aVIXa2Zm1hk9E6QknQh8B/hORHy3iU0rz0zlh6VPB74ITCE3wi/L+1BELGilrp20/mZbstFWH+h2NczMhkRPBClJ/wCcBFwP/FLSR3KrX42IeyRNAo4Dfg4sBNYD9gUOB66KiDtz28wEZgEXShpHet7qUGDnbJue4VuAZjac9USQIrV4APbOlryFwBbAM9m/TwE2JN3eewj4JvD9/AYREZKmAqdnSx+wgPRw73Vtr30H+RagmQ1nPRGkImJyA3keBT7VRJlLgSOzpaf5FqCZDVc9+zCvmZkNfw5SZmZWWg5SZmZWWg5SZmZWWg5SZmZWWg5SZmZWWj0xBN2a4wd8zWy4cJAahvyAr5kNFw5Sw5Qf8DWz4cB9UmZmVloOUmZmVloOUmZmVlruk7KaXnnlFebOnbtGukcKmtlQcJCymubOnctR582gb/yEVWkeKWhmQ8VByurqGz/BIwXNrCscpEYo38Yzs17gIDVC+TaemfUCB6kRrJO38dxSM7N2cJCyjuh0S22wQdDB06y3OEhZx3SypTbYIOjbnGZJr/zB5iA1QhTfjD5v3jxWruztZ7kHGwQ9WtGsd/5gc5AaIYpvRn/q7tvpe8/EQZVVbSqQsv31ZWb19cIfbA5SI0j+zehLnnp00OUUA97ihQ/yxY/N4/3vf/+qPMOhpWZm3ecgZatUayENFGyKAe+8Xy5Ybf6qVlpqZmYVDlK2SrXJEhsNNsX5q1ppqZmZVThI2Wq6HWx6ZcTRQHqp/r1UVxu5RnSQkjQaOB04AOgD7gdOiYhru1mv4ara7URY/aLYKyOOBtJL9e+lutrINaKDFDAd2AE4BngCOAyYLmlKRMzsZsWGo2q3E6tdFHthxFEtvVT/Yl09ctPKZsQGKUl7AXsA0yJiepY2C9gSOBtwkOqA4u3ERjQ6oKORllo7Vbtd1uujGot/SIzUlpVvhZbHiA1SwH5AP3BNJSEiQtLFwAWSto2IBV2rna3S6ICOavmKw+NbCW7FC9e8efP48R2P89a3b1WzXkOt1QtsvT8kRsIFvJ23Qtv5fXXju+/2+VZEdHwnZSTpP0lx6aOF9A8Dc4ADI+LKwroldYodAzBmzJim67NixQqWv/oaGvXGBXTlitcBMWqttar+u5W0spQVK1ey9ptGsVaWtmLFCv7y+sqa30OzaRGBNCrb3wo0aq2aebKMvGXtN6+qV6Vur/7lNahTVr1j7LRiPbNKVD2eet91tbo3Wn4vq/bdDPY8tvP7andZjRxjo/ts5dz39/dDuh6vcRtiJLekxgEPV0lfnFs/GNHf3790ENtVIlt/ccXK12v/u5W0MpS1/DWgxvG3Wj5A/k+xWLmibh6A5curZCpuM0BZAxxjLXWPv1WNHE9Fvv4N1L3p8qvo+PG3Q6PfRUNlrfl9Dfo7aPG7X72soTnfRRsAK6utGMlBCta8LtVcFxF9napIpZXWyX2UmY/fxw8j9/jB38FAereHt3WLqN5aGpt9Lq6yzszMhtBIDlL3A+/Vah0RAGyffc4f4vqYmVnBSA5S00kP8E4ppB8CPOSRfWZm3TeS+6RmArOACyWNIz3MeyiwM7BvNytmZmbJiA1S2TNRU0mvRTqd1KpaQHq497ouVs3MzDIj9jmpshnpI3t8/D5+GLnHD/4OBjKS+6TMzKzk3JIyM7PSckvKzMxKy0HKzMxKy0HKzMxKy0GqjSSNlnSupGckLZd0l6R9Gtz23ZJmSOqXtEzSTEnbDpD37yU9LOlVSY9JOqbKmzOG3GCPX9LfSLpW0sJsu0eycjaqkjcGWL7SmaNqXAvHf9IAx/TsAPmH2/l/ssZ5fbCQt8znf7yk70m6Q9KLWb0mN7F9z18DOiIivLRpAW4hvRPwi8BuwCXACmCvOtttDPwR+D0wFdibNF3In4HxhbwnZGWeAkwGjgdeA87o4eN/Gvh34HPALsBXgWeBJ4G+Qt4ALgc+Ulg27uHjPyk7rj0Kx7RDlbzD8fz/dZXz+aXsOzmjkLfM538y8BxwE2meugAmN7jtsLgGdOR77XYFhssC7JX9Uu6XSxNwB/BAnW3PApYDb8uljQOWAj8opC0HvlfY/rTsl3R8q8fRpeNf4wKTBasAvl5ID+Ccbp/vNh9/JUj11ck3LM//AOWdm5X3nl44/1ndRuV+ntpkkOr5a0CnluHZPOyOqjP9AhcD2wzUbM9te0tE/DG37SLgOmBaLt8ngHWyMvMuIr09pKFbix0y6OOPiOeqJFemAh3fzkp2UCvnv1HD8vwXSVqb1Kq+IyKqzflWShFRdT6kBg2Ha0BHOEi1z3bAgiq/qPfm1q9B0rrAu6n+1vV7gY0lbZwrI0hvcF8lIh4h/XVVdR9DZFDHX8Nu2We17+WQrM/jFUm/kfSZJsvuhHYc/wOSVmR9Oj/Mnff8PkbC+Z9KajH8eID1ZTz/gzaMrgEd4SDVPuOoPgdVvZl+30q6LdLItuOAlyPi1Sp5X6ixj6Ew2ONfg6SxpNs9jwBXFlb/FDgS+DjpjfXLgSskfaPZCrdZK8f/GKlf4XBgT+A84CBgjqS3FvYx7M8/cATwImueeyjv+W/FcLkGdMSIfcFshzQ10+8gt21lH53Wct0k/RUwgzT55MeK/xkj4vOF/D8DZgOnSrogIpY3U+E2G9TxR8SlhaRbJc0Bbgb+Dji11X0MkXac//GkQP2TiHhpjULKff5bNRyuAW3nllT7DHam3xdIv1iNbLsIWE/SW6rkfWuNfQyFlmc6zm57XEsa7bVXRNxbZ5NKP8C/A6Pp7q2Ots70HBG3AM8AOxX2MWzPf+Yw0nVpoFt9qynR+W/FcLkGdISDVPsMaqbf7C+/x6n+H2x74M+5gQX3k24LvC+fSdIEYN2B9jFEWprpWNI6pE73nYC9I+L/NbHvyj5b6bhuVSdmeh7F6sc0bM8/gCSRgtSDPXj+B20YXQM6wkGqfVqZ6Xc6sKekTSsJWb/MFODqXL4bgVeBLxS2PxR4nTQSqFsGffzZX4UzgEnAvhHx60Z3ml0UDwaWUehMHmJtnelZ0seBTUjPylQMy/OfswtpAEFDrSgo1flv1XC4BnRGt8fAD5eF9NfNrcDzpI7fXUnDQlcCU3L5ZpONzs2lbUJ6ePVu0qzAnwL+k9S0f0ch74mkX8aTSP+pjwP+Any3h4//OtLtjpNZ8yHNd+fyfQv4IfBZ0kOMBwG/zrb9Wg8f/z3A0aRnjfbMzvEy0sCRvuF+/nPrLiE967PJAOtLe/5zdfx0tpyZ1evE7N+frPM70PPXgI59p92uwHBagA2A/5P9sr2S/cJNLeSp+p8U2Ip0u2spaWTTjcD7quQTcFR2AXuVNO39P5J7kLDXjj/7zzzQclEu3xTgdtJT+K8BS4Bf5S+CPXr8l2Xn86XsYvMY8K/A2JFw/rP09bPjv6ZG+aU+/3V+l59s4Dvo+WtAJxbPJ2VmZqXlPikzMystBykzMystBykzMystBykzMystBykzMystBykzM1ul1RmGq5Q30GzKIemmets7SJkNsWy69Nn10sy6ZALpgekXSc+htWqnKsuJ2boZ9Tb2W9DNelj2F+5k0my1S7pZFxs2bouIjQEkTaXFiRQjYk4xTdLxpGlWLqu3vVtSZuWwNWmOpGZNJv1V2tfOytjIFQ3OMCxplKSjJd2XTUD5vKRL8u8fHGC7TYBPAj+PiP56+3GQMiuBiHg1Iv7S7XqYNeEnwD+TpteZQnq34q7AbEnr1djuUNJdvIZeJOwgZdYhkt4u6UpJ/ZKWSrpO0rsHyFutn+qjkm6U9Gz2l+rTkmZK+ki2/iLeuLf/RK4z+qRs/dsknS3p95JeyMpYIOlYSWsV9nVYtu1ukr4l6TFJr0p6WNKhA9R5V0k3SFqUlf24pAslbVjId2DWCb9M0svZlO+fHsRXaiUh6aOkN9x/MyL+KSJuiYiLSMHqPaQpVwZyOGlqktmN7Mt9UmYdIKkPuA14O/BvwALSG6tnkeb9qbf91sAtpJe1fg/4E7Ap8N+B95Om8Dif9FLX/UhvUX8+27wyWeR/A6aRpoF4DHgz6TbLGcCWwJer7Pr0rH7nk15e+lXgIkmPRsSdufp9GfgB8HT2uRB4B+kiNb5SF0mnAv8E3AR8h/RW9P2AqyQdGRH/t953YaW0F+lcXi4pH0fmk34ndgHWOLdZcNsG+E40+uLYbr/h1ouX4biQLvYBHF5IPydLn11IfzKfBvx9lu9DdfZzUpZviyrr1oX0EulC+qXACmCzXNphWTn3AGvn0jcnBavLcmnjs7QFFKYSydaPyj53yMo8vUqeGaS3fa/f7XPlpebv19TsHE4upP+Q2rMX3DpAeT/KfvfGN1oH3+4z64yppNbPJYX0MxvcvtKhvG82a3HTImJ5ZFcGSWtLGpvdivsF6Vb/B6tsdl7k+sYi4mngYdI0EhUHAGsDJ0eVEYXxRsf7waQL1sWSNswvpH6M9UnDka33PE9qSX0UmFhl+Wpxg6yf6jPAzRHxVKM78u0+s87YEpgbESvyiRHxjKQlDWx/OfB54HjgaElzSMHl8ohY2EgFstswx5H6DiaQ5iHKe2uVzR6vkrYIeGfu35WAdU+dKrw32+eDNfJsUqcMK6cbSb9bm0bE9Aa3OYD0h0nDMy+Dg5RZJw10z70YLNbcMOJV0nTiHwL+B/Ax4BTgJEmfa/DC8L+BrwNXAKcBz5EmC9yB1KKrdidlRZW0Yp0rP9frU1CW55M1yu3lKd+HrdzAlonZ5y5ZC/iliLgxIm7LBu5cIun7wB2kiS43J43wmxkRPysUewTpD55rmqmLg5RZZzwOvEfSWvnWlKTNgDGNFhIRvwV+m237dlLr5VTSYAioHSi+QHow86B8oqQJje5/AA9ln39Nmh12II8AnwD+f0Q80OI+bWhdVfj3SdnnQmCL7OcjSAN4vkSaKXgladDEr3lj8A4A2ajWScC50eSjFu6TMuuMa0i3sg4ppB/byMbFYdyZp0hTp4/Npb2YfY5dMzsrKLTasn6BoxupQw0/I01zf6KkDYorJVX2eWn2eXpxyHuWb+MW62EdEhEaYNkilyci4vyI+GBE/FVEjI6IrSPibyPi4UJ5j2Xbf6PZurglZdYZZwGfA34oaUfSba3JpIECz9fYruIESR8HrgeeIAWbKaThu2fl8lVeOXOmpJ+SbrnMj4j5pGDyZUlXAL8kBc3KLZdBi4inJB1FGmJ8n6RLSH9hbw7sm+3j9xExV9KJwMnA7yVdBfwR2AzYkTSMee1W6mLDn4OUWQdExAuSJpH6hQ4hBZnZpPv1jby0cwbpYv4ZUnBZTrp99iXgwtx+7pR0LPAV0rDgN5GCwnzgm8CyrIx9gT8AFwBzSUGrleP7gaTHgG+Thsu/hRSAfpXtp5LvFEm/y/IcBaxH6hubDzT9V7WNPMpGqJqZmZWO+6TMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0HKTMzKy0/gtVISd82X68fAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.set_context(\"talk\")\n",
    "a = sns.histplot(data= gene_pairs_eval_distance, x = \"distance\") "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
